Migrate domain v1 primitives to v2

NingLin-P opened the pull request:
close #1676 

This migrate domain v1 primitives to v2, including changes of:
- Construct the domain v2 `Bundle` and `ExecutionReceipt` in the domain client
- Bring back the domain runtime api like `head_receipt_number`, `oldest_receipt_number`
- Remove all the v1 usage
- Enforce the domain config's domain block limit when constructing bundle
- Revise the `extract_signer` runtime API to fix an issue that may filter out all the extrinsic when building domain block

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Misc revises of the domain v2 primitives

This commit include changes of:
- remove BundleHeader::operator_id as it is duplicated with ProofOfElection::operator_id
- add ExecutionReceipt::domain_block_hash which is needed in the fruad proof
- replace Bundle::execution_trace to ExecutionReceipt::execution_trace to store trace on chain
- add some misc helper functions

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Collect extrinsics_roots during preprocessing consensus block

It will be used later to construct the v2 ExecutionReceipt

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Replace all domain v1 primitives usage with v2 in the domain-client-operator

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Replace all domain v1 primitives usage with v2 in other misc places

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Enforce the domain config's domain block limit when construcing bundle

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add missing domain runtime api and use in the domain client

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Revise the extract_signer runtime api implementation

Calling Checkable::check before initialize_block will result in AncientBirthBlock
error due to some check in CheckMortality::additional_signed, and cause all the
extrinsics being filtered out

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Run domain bootstrapper and construct proper chain spec for the domain test

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add test test_domain_block_deriving_from_multiple_bundles

This test along with test_domain_tx_propagate is disabled due to they are using
Call::transfer internally but the evm genesis config is empty and there is no
initial balance in the testing account thus these test will failed, leavea a
TODO for now will fix in the future.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Retire submit_bundle v1

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Retire domain v1 primitives

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on domains/client/domain-operator/src/bundle_processor.rs:
```diff
@@ -243,17 +241,9 @@ where
         assert!(
             domain_block_result.header_number > head_receipt_number,
-            "Consensus chain number must larger than execution chain number by at least 1"
+            "Domain chain number must larger than execution chain number by at least 1"
```
This change is incorrect, it describes the relationship between the new domain block number derived from the new consensus block and the domain block seen by that consensus block. Furthermore, the domain chain and execution chain are the same thing, they are mostly exachangable.

NingLin-P replied:
Sorry, I don't get it. This assertion is about comparing the domain block number with the head receipt number and since the domain block is not mapping to the consensus block one by one anymore, it is weird to mention the consensus chain here IMO.

liuchengxu replied:
The mapping of consensus block number and domain block number is no longer 1:1 but the assertion still remains valid. My main point is using the domain chain number and execution chain number at the same time is sort of confusing in this context as they are usually considered the same thing unless you specifically say the execution chain number refers to the receipt chain/block tree maintained on the consensus chain. The consensus chain was mentioned to not make it confusing.

NingLin-P replied:
Okay, will replace execution chain with receipt chain.

NingLin-P committed:
Box the fraud_proof parameter of the submit_fraud_proof call

The __construct_runtime_integrity_test::runtime_integrity_tests test enforce the pallet
call enum size should be smaller than 1024 bytes, the submit_fraud_proof call break this
limit because the v2 domain header contains a receipt and the bundle equivocation proof
contains 2 headers.

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/pallet-domains/src/tests.rs:
```diff
@@ -186,11 +186,17 @@ pub(crate) fn create_dummy_receipt(
         consensus_block_hash,
         block_extrinsics_roots,
         final_state_root: Default::default(),
+        execution_trace: if block_number == 0 {
```
This change violates the spec, it shouldn't be part of ExecutionReceipt as we shall only store the trace root instead of the full trace on the chain eventually.

NingLin-P replied:
Oh, I thought we still needed to store the full trace on chain for the fraud proof (i.e. find the mismatch index of the trace), let me know if I'm wrong regarding to the fraud proof v2 design.

liuchengxu replied:
By eventually, I meant the far future, not v2 :P, so just follow the spec if no stronger reasons.

NingLin-P replied:
But we need the trace in the near future (i.e. fraud proof v2)

liuchengxu replied:
That's right. The spec already defines it as part of the bundle instead of the receipt and there is a reason for that, you can't just change the code without updating the spec.

NingLin-P replied:
Okay, the spec is also updated, let's move the disscution there.

liuchengxu commented on crates/sp-domains/src/lib.rs:
```diff
@@ -155,23 +154,28 @@ impl DomainId {
+pub struct BundleHeader<Number, Hash, DomainNumber, DomainHash, Balance> {
+    /// The consensus chain's best block number when the bundle is created. Used for detect stale
+    /// bundle and prevent attacker from reusing them to occupy the block space without cost.
     pub consensus_block_number: Number,
```
This needs to be added to the spec.

liuchengxu reviewed (commented):
A few minor comments. I believe the next step is to complete the bundle verification in terms of the new fields in v2.

vedhavyas commented on crates/sp-domains/src/v2.rs:
```diff
@@ -125,11 +130,42 @@ impl<
+    Bundle<Extrinsic, Number, Hash, DomainNumber, DomainHash, Balance>
+{
+    /// Convert a bundle with generic extrinsic to a bundle with opaque extrinsic.
+    pub fn into_opaque_bundle(
```
`From` Opaque Bundle would be more appropriate

NingLin-P replied:
We can't do that because `OpaqueBundle` is a type alias of `Bundle<OpaqueExtrinsic, ..>` thus they essentially same type.

vedhavyas commented on domains/client/domain-operator/src/domain_bundle_proposer.rs:
```diff
@@ -114,6 +116,28 @@ where
+                    })?;
+                let next_estimated_bundle_weight =
+                    estimated_bundle_weight.saturating_add(tx_weight);
+                if next_estimated_bundle_weight.any_gt(domain_block_limit.max_block_weight) {
```
is this weight specific comparaison? Can't we use `>` ?

NingLin-P replied:
Yes, because `Weight` is not impl `Ord/PartialOrd` I think this is why upstream provide `any_gt`

NingLin-P committed:
Apply review suggestion

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' into domain-client-v2

vedhavyas reviewed (commented):
Make sense in general.  Left some nits. Will look again in sometime

One thing I have noticed is that bunch of these smaller commits would actually make sense when they are merged together. Commits that just adds tests could be part of the commit that introduced the change. I recommend re-reading our contribution guidelines
